### PR TITLE
Migrate all checkbox inputs to the `CheckBoxInput` component

### LINKED ISSUE #16768: Migrate Checkbox components to the new CheckboxInputField component
### Problem to solve

In https://github.com/gitpod-io/gitpod/pull/16545, a new `CheckboxInputField` component was introduced.

Before that, we had a basic `CheckBox` component that can now be replaced.

### Proposal

To help avoid technical debt or inconsistent usage, we could migrate all checkbox input to the new component and delete the previous iteration.

See also https://github.com/gitpod-io/gitpod/issues/16511#issuecomment-1461835209. Cc @Devansu-Yadav

### PR DESCRIPTION
**Pull request by Devansu-Yadav**
## Description

- Migrating to the `CheckBoxInput` component instead of `CheckBox` component for consistency.

## Related Issue(s)

Fixes #16768

## Release Notes

```release-note
Migrate all checkbox inputs to the new CheckBoxInput component for consistency
```

## Build Options:

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-ee-license
- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Comment by Devansu-Yadav** [comment]
@gtsiolis Any updates on whether the UX for the changes made in this PR is good to go or not?

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/dashboard/src/components/forms/CheckboxInputField.tsx [hunk 1]
```diff
@@ -70,7 +70,7 @@ export const CheckboxInput: FC<CheckboxInputProps> = ({
             <div className="flex flex-col">
                 <span
                     className={classNames(
-                        "text-sm font-semibold cursor-pointer",
+                        "text-md font-semibold cursor-pointer",
```
Comment (gtsiolis) [inline components/dashboard/src/components/forms/CheckboxInputField.tsx:73]: **question**: Curious what led to this change?

| BEFORE | AFTER |
|-|-|
| check-before | check-after |

File: components/dashboard/src/components/forms/CheckboxInputField.tsx [hunk 2]
```diff
@@ -56,7 +56,7 @@ export const CheckboxInput: FC<CheckboxInputProps> = ({
             <input
                 type="checkbox"
                 className={classNames(
-                    "h-4 w-4 mt-0.5 rounded cursor-pointer border-2 dark:filter-invert",
+                    "h-4 w-4 mt-1 rounded cursor-pointer border-2 dark:filter-invert",
```
Comment (gtsiolis) [inline components/dashboard/src/components/forms/CheckboxInputField.tsx:59]: **question**: Is this change needed?

File: components/dashboard/src/components/forms/CheckboxInputField.tsx [hunk 3]
```diff
@@ -28,8 +28,8 @@ type CheckboxInputProps = {
     value: string;
     checked: boolean;
     disabled?: boolean;
-    label: string;
-    hint?: string;
+    label: string | React.ReactNode;
+    hint?: string | React.ReactNode;
```
Comment (gtsiolis) [inline components/dashboard/src/components/forms/CheckboxInputField.tsx:32]: **thought:** Not sure about this change, but let me loop in @selfcontained who initially added this component.

File: components/dashboard/src/components/forms/CheckboxInputField.tsx (hunk 2 above)
Comment (gtsiolis) [inline components/dashboard/src/components/forms/CheckboxInputField.tsx:59]: **issue(non-blocking)**: This is out of the scope of this PR, but the checkbox is using some unnecessary shades of gray, especially on the clicked state, see `2`. Cc @selfcontained 

| 1 | 2 | 3 | 4 |
|-|-|-|-|
| Frame 1374 | Frame 1375 | Frame 1376 | Frame 1377 |

File: components/dashboard/src/admin/UserDetail.tsx [hunk 4]
```diff
@@ -441,12 +449,20 @@ export default function UserDetail(p: { user: User }) {
...
             >
-                <p>Edit user permissions by adding or removing roles for this user.</p>
-                <div className="flex flex-col">
-                    {rop.map((e) => (
-                        <CheckBox key={e.title} title={e.title} desc="" checked={!!e.checked} onChange={e.onClick} />
+                <CheckboxInputField
+                    label="Edit user permissions by adding or removing roles for this user."
+                    className="mt-0"
+                >
+                    {flags.map((e) => (
```
Comment (gtsiolis) [inline components/dashboard/src/admin/UserDetail.tsx:456]: **thought**: Highlighting this part for other reviewers. 🏓

**Review by gtsiolis — CHANGES_REQUESTED** [review]
Thanks for your patience and pushing this forward, @Devansu-Yadav! 🌮 🌮 

UX LGTM! 🏁 

1️⃣ Left some minor comments below.

2️⃣ There's one more checkbox that was added recently, posting the lines of code here in case you'd like to address this change, too.

https://github.com/gitpod-io/gitpod/blob/58d838142873352ce3ce233f2946218ef75a5375/components/dashboard/src/teams/TeamSettings.tsx#L149-L161

3️⃣ Once all Checkbox components have been replaced, we can remove that file and component. It should be fine to do this separately if you prefer. Your call. 🏓 

---

@selfcontained could you take a diagonal look at the code changes here? 🏀

File: components/dashboard/src/components/forms/CheckboxInputField.tsx (hunk 1 above)
Comment (Devansu-Yadav) [reply components/dashboard/src/components/forms/CheckboxInputField.tsx:73]: @gtsiolis This change is made to maintain style consistency across all the `<CheckboxInput />` components used on different dashboard pages. The small font looks good for UX on the Onboarding flow page but doesn't look quite good on other pages that initially used the old `<CheckBox />` component (which uses the medium-sized font).

Comment (gtsiolis) [reply components/dashboard/src/components/forms/CheckboxInputField.tsx:73]: Ah, good point, @Devansu-Yadav! Yeah, the smaller font size on the onboarding was probably changed because of the many options there which ideally would be in a disclosure or a combobox dropdown. Sounds good to leave this change as is. Cc @selfcontained

File: components/dashboard/src/components/forms/CheckboxInputField.tsx (hunk 2 above)
Comment (Devansu-Yadav) [reply components/dashboard/src/components/forms/CheckboxInputField.tsx:59]: @gtsiolis As I mentioned in the review comment above, if we increase the font size for the checkbox text label from small-sized font to medium-sized font, we need to position the checkbox square down a bit to match the checkbox label text for better UX.

See below to understand this change better -
| Checkbox with `mt-0.5` for medium-sized font | Checkbox with `mt-1` for medium-sized font |
|--------|--------|
| Dashboard — Gitpod - Google Chrome 31-03-2023 12_37_12 (2) | Dashboard — Gitpod - Google Chrome 31-03-2023 12_39_29 (2) |

Comment (gtsiolis) [reply components/dashboard/src/components/forms/CheckboxInputField.tsx:59]: Thanks for clarifying, @Devansu-Yadav! Yeah, makes sense. LGTM.

File: components/dashboard/src/components/forms/CheckboxInputField.tsx (hunk 3 above)
Comment (Devansu-Yadav) [reply components/dashboard/src/components/forms/CheckboxInputField.tsx:32]: This change is made to accommodate for the usage of `<CheckBox />` component in certain places (like [`ProjectSettings`](https://github.com/gitpod-io/gitpod/pull/16813/files#diff-5d008ce9669cd9e22186bc906fca1d6263f6cdb81cd3e4e2a099a71ab48c6083L129-L138)) where React nodes are passed as props to specific attributes instead of text.

**Comment by Devansu-Yadav** [comment]
> 💡 **tip**: I've deployed up a preview environment in https://migrate-ch7cf88b5c76.preview.gitpod-dev.com/workspaces.
Thanks for this!

> 
> 1️⃣ Left some minor comments below.
Sure, I have left a comment to explain those changes :)

> 
> 2️⃣ There's one more checkbox that was added recently, posting the lines of code here in case you'd like to address this change, too.
Sure, I have pushed a commit for this.

> 
> https://github.com/gitpod-io/gitpod/blob/58d838142873352ce3ce233f2946218ef75a5375/components/dashboard/src/teams/TeamSettings.tsx#L149-L161
> 
> 3️⃣ Once all Checkbox components have been replaced, we can remove that file and component. It should be fine to do this separately if you prefer. Your call. 🏓

@gtsiolis Sure, I would also prefer to remove the old `CheckBox` component files in a separate PR.

**Comment by gtsiolis** [comment]
> I would also prefer to remove the old CheckBox component files in a separate PR.

DEAL—Sounds great, @Devansu-Yadav! No need to do this here. 🍊

File: components/dashboard/src/components/forms/CheckboxInputField.tsx (hunk 2 above)
Comment (gtsiolis) [reply components/dashboard/src/components/forms/CheckboxInputField.tsx:59]: @Devansu-Yadav Ideally, this would not have the darker border on steps 2 and 4. If this is trivial, could you include this here? Otherwise, it should be fine to address separately. Your call. 🏓

Comment (Devansu-Yadav) [reply components/dashboard/src/components/forms/CheckboxInputField.tsx:59]: @gtsiolis Sure, it's a trivial change, and I'll address it. Does this border on the checkbox square look fine in dark mode or do I need to address it as well?

Comment (gtsiolis) [reply components/dashboard/src/components/forms/CheckboxInputField.tsx:59]: Thanks @Devansu-Yadav! On the dark theme it's less noticeable but it also needs an update, see the different border color on :b:, not the focus rings. 🌔
 
| :a: | 🅱️  |
|-|-|
| Frame 1369 | Frame 1370 |

Comment (Devansu-Yadav) [reply components/dashboard/src/components/forms/CheckboxInputField.tsx:59]: @gtsiolis I have pushed a commit to address both of these.

Comment (gtsiolis) [reply components/dashboard/src/components/forms/CheckboxInputField.tsx:59]: @Devansu-Yadav Perfection. ✨ 

perfection

**Review by gtsiolis — APPROVED** [review]
UX LGTM. 🌮 🌮 

Let's wait for someone from the `@gitpod-io/engineering-webapp` team to take a closer look at the code changes. 🏓

**Review by svenefftinge — APPROVED** [review]
Code LGTM

**Comment by gtsiolis** [comment]
Thanks for taking a look, @svenefftinge. 🌮 🌮 

Removing hold and starting a build from the workspace.

/unhold

File: components/dashboard/src/components/forms/CheckboxInputField.tsx (hunk 3 above)
Comment (selfcontained) [reply components/dashboard/src/components/forms/CheckboxInputField.tsx:32]: Makes sense, thanks! We can simplify it to just the `React.ReactNode` type as that allows for the `string` type as well.

File: components/dashboard/src/components/forms/CheckboxInputField.tsx (hunk 1 above)
Comment (selfcontained) [reply components/dashboard/src/components/forms/CheckboxInputField.tsx:73]: I'd prefer we leave it as `text-sm` for now so we don't effect the onboarding UI with this change. Other occurrences of checkboxes shifting to a smaller font feel preferred imo to bumping up the vertical height on that already quite long onboarding UI w/ all the questions. We can iterate and explore bumping up the size in it's own PR where we also address the onboarding UI to keep text smaller there if we want.

File: components/dashboard/src/admin/BlockedRepositories.tsx [hunk 5]
```diff
@@ -299,17 +299,21 @@ function Details(props: {
...
-                onChange={(v) => {
-                    if (!!props.update) {
-                        props.update({ blockUser: v.target.checked });
-                    }
-                }}
-            />
+
+            <CheckboxInputContainer>
+                <CheckboxInput
+                    value="Block Users"
```
[the remark this answers is not in the document]
Comment (Devansu-Yadav) [reply components/dashboard/src/admin/BlockedRepositories.tsx:305]: @selfcontained Sure, I'll make the change. Is there anywhere else where I should drop the `value` prop?

**Comment by Devansu-Yadav** [comment]
> Thank you so much for helping with these changes 😄 I left a few pieces of feedback that would great to address as part of this before we merge it.

@selfcontained I have addressed almost all the changes you suggested. Do let me know if I missed something 😄

File: components/dashboard/src/admin/BlockedRepositories.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (selfcontained) [reply components/dashboard/src/admin/BlockedRepositories.tsx]: I think you should be safe to drop the `value` prop from all the new places in this PR where we've swapped to the new component if they weren't using it before (I think it's all of them).

**Comment by selfcontained** [comment]
Thanks @Devansu-Yadav for all your help here 😄 

I'm going to update this PR to merge into a non-`main` branch in this repo so that we can work around some things with our build pipeline that we're working through still. I'll go ahead and make the remaining adjustments (removing `value` from a few more spots), and then we'll get this into `main`.

**Comment by Devansu-Yadav** [comment]
> Update: New PR w/ some tweaks [here](https://github.com/gitpod-io/gitpod/pull/17171).

@selfcontained Just went through this PR and the code looks a whole lot better now! Thanks for the tweaks 🙌
